packages/expo-updates/android/src/main/java/expo/modules/updates/loader/FileDownloader.kt

```diff
@@ -110,6 +174,188 @@ class FileDownloader(
     }
   }
 
+  private fun isPatchResponse(responseBody: ResponseBody): Boolean {
+    val mediaType: MediaType? = responseBody.contentType()
+    return mediaType?.let { it.type == "application" && it.subtype == "vnd.bsdiff" } ?: false
+  }
+
+  @Throws(IOException::class)
+  private fun applyPatchAndVerify(
+    asset: AssetEntity,
+    request: Request,
+    responseBody: ResponseBody,
+    destination: File,
+    updatesDirectory: File,
+    expectedBase64URLEncodedSHA256Hash: String?
+  ): ByteArray {
+    val launchAssetContext = prepareAssetForDiff(
+      asset,
+      request,
+      responseBody,
+      updatesDirectory
+    )
+
+    val requestedUpdateId = request.header(EXPO_REQUESTED_UPDATE_ID_HEADER)
+
+    return responseBody.use { body ->
+      applyHermesDiff(
+        baseFile = launchAssetContext.baseFile,
+        diffBody = body,
+        destination = destination,
+        expectedBase64URLEncodedSHA256Hash = expectedBase64URLEncodedSHA256Hash,
+        asset = asset,
+        requestedUpdateId = requestedUpdateId
+      )
+    }
+  }
+
+  private data class LaunchAssetContext(val baseFile: File)
+
+  private fun prepareAssetForDiff(
+    asset: AssetEntity,
+    request: Request,
+    responseBody: ResponseBody,
+    updatesDirectory: File
+  ): LaunchAssetContext {
+    if (!asset.isLaunchAsset) {
+      failWithClosedBody(responseBody, "Received Hermes diff for non-launch asset ${asset.key}")
+    }
+
+    val currentUpdateIdHeader = request.header(EXPO_CURRENT_UPDATE_ID_HEADER)
+      ?: failWithClosedBody(
+        responseBody,
+        "Cannot apply Hermes diff without $EXPO_CURRENT_UPDATE_ID_HEADER header"
+      )
+
+    val db = database
+      ?: failWithClosedBody(responseBody, "Cannot apply Hermes diff without database access")
+
+    val currentUpdateId = try {
+      UUID.fromString(currentUpdateIdHeader)
+    } catch (e: IllegalArgumentException) {
+      failWithClosedBody(
+        responseBody,
+        "Invalid $EXPO_CURRENT_UPDATE_ID_HEADER header: $currentUpdateIdHeader",
+        e
+      )
+    }
+
+    val launchAssetEntity = db.updateDao().loadLaunchAssetForUpdate(currentUpdateId)
+      ?: failWithClosedBody(
+        responseBody,
+        "Launch asset not found for current update $currentUpdateIdHeader"
+      )
+
+    val launchAssetRelativePath = launchAssetEntity.relativePath ?: failWithClosedBody(
+      responseBody,
+      "Launch asset for update $currentUpdateIdHeader is missing a relative path"
+    )
+
+    val baseFile = File(updatesDirectory, launchAssetRelativePath)
+    if (!baseFile.exists()) {
+      failWithClosedBody(responseBody, "Base asset $baseFile is missing; cannot apply Hermes diff")
```

Comment:
```suggestion
      failWithClosedBody(responseBody, "Base asset $baseFile is missing; cannot apply patch")
```

(or "diff", not sure which is better)